### PR TITLE
Core/Spells: Add SpellMissInfo argument to BeforeHit hooks.

### PR DESCRIPTION
**Pull request by Temppus**
**Changes proposed**:
- Title said it all.
  **Target branch(es)**: 335/6x

**Issues addressed**: Partially solve one problem with #17581.

**Tests performed**: Build OK+ tested in game.

### CODE REVIEW

**Comment by Shauren** [comment]
What about all of these spells?
```
~/src/TrinityCore/src$ grep -r -n BeforeHit .
./server/scripts/Northrend/Ulduar/Ulduar/boss_flame_leviathan.cpp:1562:                BeforeHit += SpellHitFn(spell_auto_repair_SpellScript::CheckCooldownForTarget);
./server/scripts/Northrend/FrozenHalls/PitOfSaron/boss_forgemaster_garfrost.cpp:308:                BeforeHit += SpellHitFn(spell_garfrost_permafrost_SpellScript::PreventHitByLoS);
./server/scripts/Northrend/IcecrownCitadel/boss_blood_prince_council.cpp:1566:                BeforeHit += SpellHitFn(spell_valanar_kinetic_bomb_knockback_SpellScript::KnockIntoAir);
./server/scripts/Northrend/IcecrownCitadel/boss_blood_queen_lana_thel.cpp:588:                BeforeHit += SpellHitFn(spell_blood_queen_vampiric_bite_SpellScript::OnCast);
./server/scripts/Northrend/IcecrownCitadel/boss_icecrown_gunship_battle.cpp:2143:                BeforeHit += SpellHitFn(spell_igb_below_zero_SpellScript::RemovePassengers);
./server/scripts/Northrend/IcecrownCitadel/boss_the_lich_king.cpp:2119:                BeforeHit += SpellHitFn(spell_the_lich_king_necrotic_plague_SpellScript::CheckAura);
./server/scripts/Northrend/IcecrownCitadel/boss_the_lich_king.cpp:2938:                BeforeHit += SpellHitFn(spell_the_lich_king_restore_soul_SpellScript::RemoveAura);
./server/scripts/Spells/spell_warlock.cpp:176:                BeforeHit += SpellHitFn(spell_warl_banish_SpellScript::HandleBanish);
./server/scripts/Spells/spell_rogue.cpp:284:            void HandleBeforeHit()
./server/scripts/Spells/spell_rogue.cpp:348:                BeforeHit += SpellHitFn(spell_rog_deadly_poison_SpellScript::HandleBeforeHit);
```
Should they also trigger if the hit was not successful?

**Comment by Temppus** [comment]
Hard to say. Some of them yes, but i have no deeper knowledge if there are are some circumstances which can trigger spell to fail. This change was proposed due #17581 as  I mentioned.

I came up with some possible ways:

1.) Rewrite BeforeHit hooks to OnHit, but again too much spellscripts use it, dunno if this change will fuck u something.
2.) Create new hook (but as far as I know this is only spell that requires it).
3.) Some workaround in spellscript or in core for that Banish.

Not sure which one to choose, or if there is some other proper solution.

**Comment by Shauren** [comment]
What about adding a SpellMissInfo argument to BeforeHit hook?

**Comment by Temppus** [comment]
Yep, that seems best solution.

File: src/server/game/Spells/SpellScript.h [hunk 1]
```diff
@@ -307,8 +318,10 @@ class TC_GAME_API SpellScript : public _SpellScript
         HookList<EffectHandler> OnEffectSuccessfulDispel;
         #define SpellEffectFn(F, I, N) EffectHandlerFunction(&F, I, N)
 
-        // example: BeforeHit += SpellHitFn(class::function);
-        HookList<HitHandler> BeforeHit;
+        // example: BeforeHit += BeforeSpellHitFn(class::function);
```
Comment (Shauren) [inline src/server/game/Spells/SpellScript.h]: please document expected hook signature, just like here https://github.com/Temppus/TrinityCore/blob/bc2634a405d8fecd350fe552fa277ef3f4987717/src/server/game/Spells/SpellScript.h#L313

**Comment by Temppus** [comment]
Should be ready.

**Comment by Temppus** [comment]
Done.

**Pull request merged**